common/src/main/java/io/netty/util/concurrent/MultithreadEventExecutorGroup.java

```diff
@@ -15,49 +15,65 @@
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 /**
- * Abstract base class for {@link EventExecutorGroup} implementations that handles their tasks with multiple threads at
- * the same time.
+ * Abstract base class for {@linkplain EventExecutorGroup}s implementations that manage multiple
+ * {@linkplain EventExecutor} instances.
  */
-public abstract class MultithreadEventExecutorGroup extends AbstractEventExecutorGroup {
+public abstract class MultithreadEventExecutorGroup<T1 extends EventExecutor, T2 extends EventExecutorMetrics<T1>>
+        extends AbstractEventExecutorGroup {
 
-    private final EventExecutor[] children;
-    private final Set<EventExecutor> readonlyChildren;
-    private final AtomicInteger childIndex = new AtomicInteger();
     private final AtomicInteger terminatedChildren = new AtomicInteger();
     private final Promise<?> terminationFuture = new DefaultPromise(GlobalEventExecutor.INSTANCE);
-    private final EventExecutorChooser chooser;
+    private final EventExecutorScheduler<T1, T2> scheduler;
 
     /**
-     * @param nEventExecutors   the number of {@link EventExecutor}s that will be used by this instance.
-     *                           If {@code executor} is {@code null} this number will also be the parallelism
-     *                           requested from the default executor. It is generally advised for the number
-     *                           of {@link EventExecutor}s and the number of {@link Thread}s used by the
-     *                           {@code executor} to lie very close together.
+     * @param nEventExecutors   the number of {@linkplain EventExecutor}s that will be used by this instance.
+     *                          If {@code Executor} is {@code null} this will also be the number of threads
+     *                          requested from the {@link DefaultExecutorFactory}. It is advised for
+     *                          {@code nEventExecutors} and the number of threads used by the {@code Executor} to lie
+     *                          very close together.
      * @param executorFactory   the {@link ExecutorFactory} to use, or {@code null} if the default should be used.
-     * @param args                arguments which will passed to each {@link #newChild(Executor, Object...)} call
+     * @param scheduler         the {@link EventExecutorScheduler} that, on every call to
+     *                          {@link EventExecutorGroup#next()}, decides which {@link EventExecutor} to return.
+     * @param args              arguments which will passed to each
+     *                          {@link #newChild(Executor, EventExecutorMetrics, Object...)} call.
      */
-    protected MultithreadEventExecutorGroup(int nEventExecutors, ExecutorFactory executorFactory, Object... args) {
-        this(nEventExecutors, executorFactory == null ? null : executorFactory.newExecutor(nEventExecutors), args);
```

Comment:
Please keep the constructors without `EventExecutorScheduler` and `EventExecutorMetricsFactory` so that the subclasses can provide the constructors without them easily.